rascal-lsp/src/main/rascal/lang/rascal/lsp/IDECheckerWrapper.rsc

```
@@ -0,0 +1,96 @@
+    and calls the type checker on (transitive) dependencies first.
+
+    This function must only be used in an IDE context.
+}
+list[ModuleMessages] checkFile(loc l, node(loc file) getParseTree, PathConfig(loc file) getPathConfig) {
+    checkForImports = [pt | start[Module] pt := getParseTree(l)];
+    checkedForImports = {};
+
+    rel[loc, loc] dependencies = {<root, root> | root := inferProjectRoot(l)};
+    
+    msgs = [];
+
+    while (tree <- checkForImports) {
+        currentSrc = tree.src.top;
+        currentProject = inferProjectRoot(currentSrc);
+        for (i <- tree.top.header.imports) {
+            ml = getRascalModuleLocation("<i.\module>", getPathConfig(currentProject));
+            mlProject = inferProjectRoot(ml);
+            dependencies += <currentProject, resolveLocation(mlProject)>;
+            if (ml.extension == "rsc" && ml notin checkedForImports, start[Module] mlpt := getParseTree(ml)) {
+                checkForImports += mlpt;
+            }
+        }
+        checkedForImports += currentSrc;
+        checkForImports -= tree;
+    }
+    modulesPerProject = classify(checkedForImports, loc(loc l) {return inferProjectRoot(l);});
+    return [*check([*modulesPerProject[project]], makeCompilerConfig(getPathConfig(project))) | project <- reverse(order(dependencies)), project in modulesPerProject];
+}
+
+RascalCompilerConfig makeCompilerConfig(PathConfig pcfg) = rascalCompilerConfig(pcfg[resources=pcfg.bin]);
+
+loc inferProjectRoot(loc member) {
+    current = member;
+    if (!isDirectory(current)) {
+        current = current.parent;
+    }
+
+    while (exists(current), isDirectory(current)) {
+        if (exists(current + "META-INF" + "RASCAL.MF")) {
+            return ensureTrailingSpaceInPath(current);
+        }
+        if (!current.parent?) {
+            return isDirectory(member) ? member : member.parent;
+        }
+        current = current.parent;
```

what if `current.parent == current`?

---

Doesn't Rascal always throw a `NoParent` instead of returning the location itself?

---

The condition in the `if` 'catches' the `NoParent`s

---

aaaah, right, my bad.